### PR TITLE
New TPR Service listing component with search

### PR DESCRIPTION
**Pull request by dragos-dumi-ibm**
More details at
[PLAENH-185](https://helsinkisolutionoffice.atlassian.net/browse/PLAENH-185)
[UHF-7656](https://helsinkisolutionoffice.atlassian.net/browse/UHF-7656)

### CODE REVIEW

File: src/Plugin/migrate/process/TprEmptyArray.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ *   source: array
+ * @endcode
+ *
+ */
+class TprEmptyArray extends ProcessPluginBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transform($value, MigrateExecutableInterface $migrate_executable, Row $row, $destination_property) {
```
Comment (tuutti) [inline src/Plugin/migrate/process/TprEmptyArray.php:32]: Missing return type (`bool`).

File: src/Plugin/views/argument/ServiceIdArgument.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * @ViewsArgument("id_or_service_id_handler")
+ */
+class ServiceIdArgument extends ArgumentPluginBase
+{
+
+  /**
+   * Build the query based upon the formula
+   */
+  public function query($group_by = FALSE) {
```
Comment (tuutti) [inline src/Plugin/views/argument/ServiceIdArgument.php:22]: Missing return type (`void`).

File: tests/src/Kernel/ServiceIdArgumentTest.php [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function setUp(): void
+  {
+    parent::setUp();
+  }
+
+
+  public function testServiceIdArgument()
```
Comment (tuutti) [inline tests/src/Kernel/ServiceIdArgumentTest.php:28]: Missing retun type (`void`).

File: tests/src/Kernel/ServiceIdArgumentTest.php [hunk 4]
```diff
@@ -0,0 +1,136 @@
+<?php
+
```
Comment (tuutti) [inline tests/src/Kernel/ServiceIdArgumentTest.php:2]: Missing `declare(strict_types = 1);`.

File: tests/modules/helfi_tpr_test/helfi_tpr_test.info.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+name: 'Service List Test Config'
+type: module
+core_version_requirement: ^8 || ^9
```
Comment (tuutti) [inline tests/modules/helfi_tpr_test/helfi_tpr_test.info.yml:3]: core_version_requirement should be `^9 || ^10`.

File: src/Entity/Service.php [hunk 6]
```diff
@@ -150,6 +150,39 @@ protected function getErrandServiceIndex(ErrandService $errand_service) {
   public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
     $fields = parent::baseFieldDefinitions($entity_type);
 
+    $fields['name_synonyms'] = BaseFieldDefinition::create('string_long')
```
Comment (tuutti) [inline src/Entity/Service.php:153]: This field should probably have an unlimited cardinality, then separate the `name_synonyms` field by comma. See Phone field (Unit entity/migration) for an example how to implement this.

File: helfi_tpr.module [hunk 7]
```diff
@@ -269,3 +269,20 @@ function helfi_tpr_tpr_errand_service_access(EntityInterface $entity, $operation
 function helfi_tpr_query_owd_relationship_alter(AlterableInterface $query) : void {
   $query->groupBy('id');
 }
+
+/**
+ * Implements hook_views_data_alter().
+ */
+function helfi_tpr_views_data_alter(&$data) {
```
Comment (tuutti) [inline helfi_tpr.module:276]: Missing return type (`void`).

**Review by tuutti — CHANGES_REQUESTED** [review]

File: src/Plugin/Field/FieldFormatter/ErrandServicesFormatter.php [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+          'relationship' => 'sibling',
+          'group' => 'group-weight',
+        ],
+      ],
+    ];
+    $channelTypes = [];
+    foreach ($this->getChannelTypes() as $channelId => $channelType) {
+      $channelTypes[$channelId] = $channelType;
+    }
+    $channelTypes['OFFICE'] = new ChannelType('OFFICE', 99);
```
[the remark this answers is not in the document]
Comment (dragos-dumi-ibm) [reply src/Plugin/Field/FieldFormatter/ErrandServicesFormatter.php:96]: @tuutti for this feature, a 'fake' channel Office is needed to be displayed if service has at least a unit. there's no link to an actual channel type. We used it as ChannelType to add it next to the others and use it in the widget.

Based on the discussion on https://helsinkisolutionoffice.atlassian.net/browse/PLAENH-185  we should list the chips/pills based on the channel types found in errand services.

File: src/Entity/Service.php (hunk 6 above)
Comment (dragos-dumi-ibm) [reply src/Entity/Service.php:153]: @tuutti yes, logically make sense to be multiple. but if we do the name synonyms multiple, then won't be able to use combine search field (one requirements was to search in name synonyms) we would have to maybe do aggregation on view which I think would overcomplicate things for this feature.

**Review by tuutti — APPROVED** [review]

**Pull request merged**